### ISSUE TITLE
"Killed Innocent" morale debuff from ambushing hostile bandit

### ISSUE THREAD

**Issue by chaoticidealism** [CleverRaven/Cataclysm-DDA#30124]
**Describe the bug**  
If you attack a hostile NPC before they "get angry", the kill will be counted as a "Killed Innocent" morale penalty for all non-psychopath/bloodthirsty PCs. This is true even if your "attack" is an attempt to disarm them, or if they are known bandits (quest targets), or if they are attempting to rob you.

**To Reproduce**  
1. Take a bandit quest from the refugee center. Go to the bandits' location.
2. Attack the bandits before getting the "Gets angry" message. Long-range weapons work; so does invisibility, or short-range weapons in darkness.
3. Observe morale level.

**Expected behavior**  
Ambushing hostile NPCs shouldn't give you that morale debuff. You knew they were going to try to kill you; you killed them first. That's very different from killing someone who, say, absent-mindedly picked up your only soldering iron but means you no harm. This also discourages the use of ambush tactics on NPCs. I would expect no morale drawbacks to killing hostile NPCs, whether or not they had been alerted to your presence and attacked you first. If this bug is kept as a feature, it might be retained for PCs with the "Pacifist" trait.

**Versions and configuration(please complete the following information):**
Literally every version for the past year.

**Comment by davidpwbrown** [comment]
I've not encountered this with bandits in the bandit tower.
Perhaps something wrong with the way the mission spawns them?

**Comment by chaoticidealism** [comment]
It happens mostly when you use an invisibility bionic, but can happen if you use a very long range weapon with improved night vision in the dark.

**Comment by mlangsdorf** [comment]
the check to see if you killed an innocent is pretty straightforward:
```
    if( killer == &g->u && ( !guaranteed_hostile() || hit_by_player ) ) {
```
and guaranteed hostile is easy:
```
bool npc::guaranteed_hostile() const
{
    return is_enemy() || ( my_fac != nullptr && my_fac->likes_u < -10 );
}
```
so unless there's some lazy evaluation of my_fac going on, which there might be, it shouldn't matter whether the NPC has seen you or not.

What's weird is that the logic that makes them go hostile when they see you in a check against guaranteed_hostile() so I don't know why it would matter if they can see you.

**Comment by Scischm** [comment]
The way I'm seeing that code for the check (in plain english) where this case is happening goes something like this?

`If ( you are the killer (true) &&  ( the enemy is not hostile (false) || you hit the npc (true))`

 couldn't that result in a case where the if statement returned true despite the fact the enemy was hostile just because you hit them?

Should that `||` statement be an `&&` statement? Since the person would have to be both non-hostile and have been hit and killed by the player?

**Comment by drachs** [comment]
Happened to me on "Kill Bandits" from the refugee center representative.  I used a cloaking bionic to slip in through the window unseen and killed them with the mono molecular blade.   Wow, they have a lot of hitpoints!   My character is apparently going to spend the next few days anguished over the fact that they've become a post human killing machine.  Those bandits saw nothing more than a gust of wind blowing through the curtains of a window they swore they closed.

**Comment by chaoticidealism** [comment]
> 
> 
> Happened to me on "Kill Bandits" from the refugee center representative. I used a cloaking bionic to slip in through the window unseen and killed them with the mono molecular blade. Wow, they have a lot of hitpoints! My character is apparently going to spend the next few days anguished over the fact that they've become a post human killing machine. Those bandits saw nothing more than a gust of wind blowing through the curtains of a window they swore they closed.

That's exactly the phenomenon I'm seeing too. Sneaking up on people, or even just being the first to land a hit, seems to cause your character to think they've killed an innocent even though the NPC was hostile to begin with and a known bandit to boot.

Check your game setup and compare it to mine--maybe we have something in common that seems to trigger the bug? In the meantime, I feel no guilt over cheating myself some Prozac.

**Comment by Maleclypse** [comment]
I am unable to reproduce this bug having tried every mentioned method above and any I could think of.

**Comment by chaoticidealism** [comment]
Can confirm it's no longer happening in my game either. Must have gotten fixed along the way while someone was dealing with some other bug.

**Comment by VenomousInc** [comment]
> ![image](https://user-images.githubusercontent.com/89065534/198377311-eb2d5af1-c77a-404b-ab5f-7586e0b03485.png) ![image](https://user-images.githubusercontent.com/89065534/198377381-6bce40c9-5de8-4673-8993-540b97d303a8.png)
> 
> I also got this de-buff after killing the 3 bandits from the "missing caravan" quest
> 
> Its worth mentioning that I was not at a distance or invisible, I crashed my van into their motorcycles/ATV and had a shootout with them. Build: 2022-10-15-0243

I just got this issue during the `Missing Caravan` Quest, they're not innocent so it doesn't make sense; especially after one of shot at me first.

**Comment by AlexanderCecile** [comment]
I have just encountered this issue on build 2023-02-20-0616. 2 NPCs, clearly marked as bandits. Taking some of their items was also considered stealing.

**Comment by AkdoolTheSlimy** [comment]
Build 2023-03-04-0252 and I'm also running into this issue with bandits from forts. I used bionic sniper's night vision bionic while it's nighttime/too dark to be seen, and I'm constantly dropped to -300ish morale due to killing innocents. They're marked as hostile even though they can't see me.

**Comment by vjek** [comment]
Just had this happen when killing thugs/bandits in 0.G ( Game Version: d6ec466 [64-bit] ), on March 12, 2023.
All NPCs were hostile, and after killing them, taking all their items was considered stealing.  
My character took a  negative 100% morale hit due to " Killed an innocent person ".

**Comment by ektoplazm** [comment]
- OS: Windows
    - OS Version: 10.0.19045.3208 (22H2)
- Game Version: 55a749a [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

Happened here too with bandits in wooden fort.

**Comment by ampersand55** [comment]
Me too, road block bandits.

**Comment by Cenness** [comment]
Still replicable in 

- OS: Windows
    - OS Version: 10.0.19045.3803 (22H2)
- Game Version: e64a948 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda]
]
